keras_cv/metrics/object_detection/box_coco_metrics.py

```diff
@@ -0,0 +1,285 @@
+# Copyright 2023 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import sys
+import types
+
+import tensorflow as tf
+import tensorflow.keras as keras
+
+import keras_cv
+from keras_cv import bounding_box
+
+
+class HidePrints:
+    """A basic internal only context manager to hide print statements."""
+
+    def __enter__(self):
+        self._original_stdout = sys.stdout
+        sys.stdout = open(os.devnull, "w")
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.stdout.close()
+        sys.stdout = self._original_stdout
+
+
+def _box_concat(b1, b2):
+    """Concatenates two bounding box batches together."""
+    if b1 is None:
+        return b2
+    if b2 is None:
+        return b1
+
+    result = {}
+    for key in ["boxes", "classes"]:
+        result[key] = tf.concat([b1[key], b2[key]], axis=0)
+
+    if "confidence" in b2:
+        result["confidence"] = tf.concat(
+            [b1["confidence"], b2["confidence"]], axis=0
+        )
+    return result
+
+
+METRIC_NAMES = [
+    "AP",
+    "AP50",
+    "AP75",
+    "APs",
+    "APm",
+    "APl",
+    "ARmax1",
+    "ARmax10",
+    "ARmax100",
+    "ARs",
+    "ARm",
+    "ARl",
+]
+
+
+class BoxCOCOMetrics(keras.metrics.Metric):
+    """BoxCOCOMetrics computes standard object deteciton metrics.
+
+    Args:
+        bounding_box_format: the bounding box format for inputs.
+        evaluate_freq: the number of steps to run before each evaluation.
+            Due to the high computational cost of metric evaluation the final
+            results are only updated once every `evaluate_freq` steps.  Higher
+            values will allow for faster training times, while lower numbers
+            allow for higher numerical precision in metric reporting.
+
+    Usage:
+    `BoxCOCOMetrics()` can be used like any standard metric with any
+    KerasCV object detection model.  Inputs to `y_true` must be KerasCV bounding
+    box dictionaries, `{"classes": classes, "boxes": boxes}`, and `y_pred` must
+    follow the same format with an additional `confidence` key.
+
+    Using this metric suite alongside a model is trivial; simply provide it to
```

Comment:
I assume using this with our OD models breaks TPU compat. If so, probably worth mentioning that for TPU users should use the COCOCallback instead.

Reply:
Right now my plan for TPU is actually to remove the COCO Callback and ship a generic PyMetric evaluator callback.  

We should not recommend use of the callback anymore, instead we should ship a generic API.

Reply:
Updated with a blurb of insight on how to evaluate it right now.